### PR TITLE
Backend rspec unit tests

### PR DESCRIPTION
**Pull request by benpony**
Rspec, FactoryBot
review model spec
review controller spec

### CODE REVIEW

File: spec/controllers/reviews_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
+require 'rails_helper'
+
+RSpec.describe Api::ReviewsController, type: :controller do
+  def get_review(stars, description)
```
Comment (ronits-wework) [inline spec/controllers/reviews_controller_spec.rb]: this should be a factory

File: spec/controllers/reviews_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  describe 'rest api' do
+    context 'index succeeds' do
+      it 'returns success' do
+        get :index
+        expect(response).to have_http_status(:success)
+      end
+
+      it 'returns an array of reviews' do
+        get :index
+        expect(JSON.parse(response.body)).is_a? Array
```
Comment (ronits-wework) [inline spec/controllers/reviews_controller_spec.rb]: my personal preference is to unite tests if I can, just because it's more efficient, but it's just my preference

File: spec/controllers/reviews_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  def get_review(stars, description)
+    {
+      user_id: 1,
+      restaurant_id: 1,
+      stars: stars,
+      description: description
+    }
+  end
+
+  describe 'rest api' do
```
Comment (ronits-wework) [inline spec/controllers/reviews_controller_spec.rb]: we usually use describe for the method and context for different cases in that method
so in this case: `describe '#index'`, and no need for context because you just have one case (success)

File: spec/controllers/reviews_controller_spec.rb (hunk 2 above)
Comment (ronits-wework) [reply spec/controllers/reviews_controller_spec.rb]: also, you don't create reviews before calling index, so this test will always receive an empty array

File: spec/controllers/reviews_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      end
+
+      it 'returns success' do
+        get :show, params: valid_params
+        expect(response).to have_http_status(:success)
+      end
+
+      it 'returns the requested review record' do
+        get :show, params: valid_params
+        expect(JSON.parse(response.body).symbolize_keys[:id]).to eq(1)
```
Comment (ronits-wework) [inline spec/controllers/reviews_controller_spec.rb]: there's no review in the DB, so how does it return anything?

File: spec/controllers/reviews_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      end
+    end
+
+    context 'create succeeds' do
+      let(:valid_params) do
+        { review: get_review(2, 'create review') }
```
Comment (ronits-wework) [inline spec/controllers/reviews_controller_spec.rb]: use a factory instead

File: spec/controllers/reviews_controller_spec.rb (hunk 1 above)
Comment (ronits-wework) [inline spec/controllers/reviews_controller_spec.rb]: I think you should rename to 'review_params', because it's not a review record, it's the params sent to create one

File: spec/controllers/reviews_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      end
+
+      it 'returns success' do
+        post :create, params: valid_params
+        expect(response).to have_http_status(:success)
+      end
+
+      it 'creates the review with the correct attributes' do
+        post :create, params: valid_params
+        expect(JSON.parse(response.body).symbolize_keys[:id].present?).to eq(true)
```
Comment (ronits-wework) [inline spec/controllers/reviews_controller_spec.rb]: also nice to know rspec magic - when there's a boolean method (and this works for any boolean method, even custom ones) like present?, you can do:
`expect(JSON.parse(response.body).symbolize_keys[:id]).to be_present`

File: spec/factories.rb [hunk 7]
```diff
@@ -0,0 +1,71 @@
+FactoryBot.define do
+  factory :user, class: User do
```
Comment (ronits-wework) [inline spec/factories.rb]: in general we have a different file for each factory

File: spec/factories.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    trait :admin do
```
Comment (ronits-wework) [inline spec/factories.rb]: I think you should use traits for special cases, not for the default case
because you don't provide anything for the default case, then the user will always have to pass the trais - so why use a trait at all?

Comment (ronits-wework) [reply spec/factories.rb]: if your trait is admin, then it should only set isAdmin inside it, and the default outside should set it to false

File: spec/models/review_spec.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
+require 'rails_helper'
+require 'date'
+
+RSpec.describe Review, :type => :model do
+  # pending "add some examples to (or delete) #{__FILE__}"
+
+  before(:all) do
+    @restaurant = FactoryBot.create(:restaurant, :taizu)
```
Comment (ronits-wework) [inline spec/models/review_spec.rb]: prefer using `let` to instance variables:
`let(:restaurant) { FactoryBot.create(:restaurant, :taizu) }`

File: spec/models/review_spec.rb [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  end
+
+  def get_review(stars)
+    create(:review, stars: stars, restaurant_id: @restaurant.id)
```
Comment (ronits-wework) [inline spec/models/review_spec.rb]: prefer using `let` to a method:
`let(:review) { create(:review, stars: stars, restaurant_id: restaurant.id) }`

Comment (ronits-wework) [reply spec/models/review_spec.rb]: you can define the number of stars in a let as well

File: spec/models/review_spec.rb [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+      it 'has a user' do
+        expect(User.find(med_review.user_id)).to be_truthy
+      end
+
+      it 'has a restaurant' do
+        expect(Restaurant.find(med_review.restaurant_id)).to be_truthy
+      end
+
+      it 'has a stars rank' do
+        expect(med_review.stars).to be_a(Numeric)
```
Comment (ronits-wework) [inline spec/models/review_spec.rb]: you can even expect it to be 3, no?

File: spec/models/review_spec.rb [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+      end
+    end
+
+    context 'no duplicate reviews' do
```
Comment (ronits-wework) [inline spec/models/review_spec.rb]: I don't understand what this is testing - is there a validation to not allow the same user to review twice in the same day?

File: spec/models/review_spec.rb [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+          }
+          expect(unique_reviews.size).to eq(reviews.size)
+        end
+      end
+    end
+  end
+
+  describe 'active record callbacks' do
+    before(:each) do
+      @reviewed_restaurant = FactoryBot.create :restaurant, :taizu
```
Comment (ronits-wework) [inline spec/models/review_spec.rb]: prefer let to instance variables

File: app/models/address.rb [hunk 14]
```diff
@@ -16,5 +16,5 @@ class Address < ApplicationRecord
 
   validates_presence_of :city, :street, :street_number
   validates_length_of :city, :street, maximum: 255
-  validates_numericality_of :street_number, :zipcode
+  validates_numericality_of :street_number, :zipcode, :allow_blank => true
```
Comment (pazaricha) [inline app/models/address.rb]: minor: `allow_blank: true`

File: spec/controllers/reviews_controller_spec.rb (hunk 1 above)
Comment (pazaricha) [reply spec/controllers/reviews_controller_spec.rb]: I second that.

File: spec/controllers/reviews_controller_spec.rb (hunk 2 above)
Comment (pazaricha) [reply spec/controllers/reviews_controller_spec.rb]: Correct, please create some reviews than test that they are indeed included in the returned array.

File: spec/controllers/reviews_controller_spec.rb (hunk 4 above)
Comment (benpony) [reply spec/controllers/reviews_controller_spec.rb]: there is a test db, but i should update to the Factory !
